src/datachain/lib/dc.py

```diff
@@ -475,25 +475,29 @@ def from_storage(
             dc.signals_schema = dc.signals_schema.mutate({f"{object_name}": file_type})
             return dc
 
+        dc = cls.from_dataset(list_ds_name, session=session, settings=settings)
+        dc.signals_schema = dc.signals_schema.mutate({f"{object_name}": file_type})
+
         if update or not list_ds_exists:
-            # disable prefetch for listing, as it pre-downloads all files
-            (
-                cls.from_records(
-                    DataChain.DEFAULT_FILE_RECORD,
-                    session=session,
-                    settings=settings,
-                    in_memory=in_memory,
-                )
-                .settings(prefetch=0)
-                .gen(
-                    list_bucket(list_uri, cache, client_config=client_config),
-                    output={f"{object_name}": File},
+
+            def lst_fn():
+                # disable prefetch for listing, as it pre-downloads all files
+                (
+                    cls.from_records(
+                        DataChain.DEFAULT_FILE_RECORD,
+                        session=session,
+                        settings=settings,
+                        in_memory=in_memory,
+                    )
+                    .settings(prefetch=0)
+                    .gen(
+                        list_bucket(list_uri, cache, client_config=client_config),
+                        output={f"{object_name}": File},
```

Comment:
If I am not mistaken:

```suggestion
                        output={f"{object_name}": file_type},
```